api/envoy/extensions/dso/v3/dso.proto

```diff
@@ -0,0 +1,38 @@
+syntax = "proto3";
+
+package envoy.extensions.dso.v3;
+
+import "xds/annotations/v3/status.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.dso.v3";
+option java_outer_classname = "DsoProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/dso/v3;dsov3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#protodoc-title: Dso]
+// Dso config.
```

Comment:
This seems unrelated to the Go extension, and is a much larger conversation. Can you remove this for now and we can have a separate conversation about this? I'm not sure why this is required for base level Go plugin support.

Edit: Or maybe it is related? TBH I don't really understand how this is to be used by someone, why it requires SO support, etc. If the idea is that the Go modules are compiled as SOs, there needs to be a large conversation about versions, versions checks, ABI, etc. Additionally, can the SO config just be part of the Go filter config so it's self contained from an API perspective.

Would it be possible to push some more documentation on how someone might actually use this? It's really not clear to me from the API. I don't think we need the full implementation right now but a bit more fleshed out usage guide with perhaps a filter example and how one would compile it would be nice as a point of discussion?

Thank you.

Reply:
> This seems unrelated to the Go extension, and is a much larger conversation. Can you remove this for now and we can have a separate conversation about this? I'm not sure why this is required for base level Go plugin support.
> 
> Edit: Or maybe it is related? TBH I don't really understand how this is to be used by someone, why it requires SO support, etc. If the idea is that the Go modules are compiled as SOs, there needs to be a large conversation about versions, versions checks, ABI, etc. Additionally, can the SO config just be part of the Go filter config so it's self contained from an API perspective.
> 
> Would it be possible to push some more documentation on how someone might actually use this? It's really not clear to me from the API. I don't think we need the full implementation right now but a bit more fleshed out usage guide with perhaps a filter example and how one would compile it would be nice as a point of discussion?
> 
> Thank you.
-----

First, developer's  Go plugins were integrated into Envoy via dynamic libraries, which can avoided rebuilding Envoy when adding new Go plugins.
By design, the SO config could be part of the Go filter config, but there is a user experience issue that requires the complete SO config for each Go plugin, which seems a bit complicated. Therefore, we separate the SO config from the Go plugin as a general API for SO management. Finally, we can refer to the specific SO via `so_id` in the go plugin config.

Take the following example:

develop go-plugin
====

Envoy developers can implement the interface of `StreamFilter` [API](https://github.com/envoyproxy/envoy/blob/9499dbb0941464528653914cb9eb31a77ec7c1da/contrib/go_plugin_api/filter.go#L5) by the GO language to achieve the effects of Envoy native filter.

Then put the GO plugin source code into the `${OUTPUT}/src/` directory with the name of the plugin for GO plugin builds. The following examples implement `limit` and `auth` GO plugins.
```
tree /home/admin/envoy/go-plugins/src/
├── auth
│   ├── factory.go
│   └── filter.go
└── limit
    ├── factory.go
    └── filter.go

````

build go-plugin
====
go_plugin_generate.sh
``` bash
#!/bin/bash
if [ $#  != 2 ]; then 
   echo "need input the go plugin name"
   exit 1
fi

PLUGINNAME=$1
OUTPUT=/home/admin/envoy/go-plugins/
PLUGINSRCDIR=${OUTPUT}/src/${PLUGINNAME}
go build --buildmode=c-shared  -v -o $OUTPUT/lib${PLUGINNAME}.so $PLUGINSRCDIR
```
Build the Go plugin so by `go_plugin_generate.sh` script,  below example the `liblimit.so` and `liblimit.so` will be generated in the `/home/admin/envoy/go-plugins/` directory.
* go_plugin_generate.sh limit
* go_plugin_generate.sh auth

config go-plugin
====

``` yaml
bootstrap_extensions:
  - name: envoy.bootstrap.dso
    typed_config:
      "@type": type.googleapis.com/envoy.extensions.dso.v3.dso
      so_id: plugin-limit
      so_path: "/home/admin/envoy/go-plugins/liblimit.so"
  - name: envoy.bootstrap.dso
    typed_config:
      "@type": type.googleapis.com/envoy.extensions.dso.v3.dso
      so_id: plugin-auth
      so_path: "/home/admin/envoy/go-plugins/libauth.so"      
static_resources:
  listeners:
    - name: ingress
      address:
        socket_address:
          protocol: TCP
          address: 0.0.0.0
          port_value: 8080
      filter_chains:
        - filters:
            - name: envoy.filters.network.http_connection_manager
            …………
                http_filters:
                  - name: envoy.filters.http.golang
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.golang.v3.Config
                      so_id: plugin-limit
                      plugine_name: limit
                      plugin_config:
                        "@type": type.googleapis.com/envoy.extensions.filters.http.golang.plugins.limit.v3.Config
                        xxx1: xx1
                        xxx2: xx2
                  - name: envoy.filters.http.header_to_metadata
                  - name: envoy.filters.http.golang
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.golang.v3.Config
                      so_id: plugin-auth
                      plugine_name: auth
                      plugin_config:
                        "@type": type.googleapis.com/envoy.extensions.filters.http.golang.plugins.auth.v3.Config
                        xxx1: xx1
                        xxx2: xx2
                  - name: envoy.filters.http.router
    - name: egress
      address:
        socket_address:
          protocol: TCP
          address: 0.0.0.0
          port_value: 8081
      filter_chains:
        - filters:
            - name: envoy.filters.network.http_connection_manager
            …………
                http_filters:
                  - name: envoy.filters.http.golang
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.golang.v3.Config
                      so_id: plugin-auth
                      plugine_name: auth
                      plugin_config:
                        "@type": type.googleapis.com/envoy.extensions.filters.http.golang.plugins.auth.v3.Config
                        xxx1: xx1
                        xxx2: xx2
                  - name: envoy.filters.http.router                  
```

Reply:
ping @mattklein123

Reply:
If we go with this approach there needs to be a lot more in here in terms of documentation, an example filter, etc. Can you please flesh this out more in terms of this PR?